### PR TITLE
Add functionality to test dependencies by changing test running order

### PR DESCRIPTION
**Pull request by epdenouden**
|Note: updated documentation|
|:---|
|There is a wiki page with more [documentation with examples](https://github.com/epdenouden/phpunit/wiki/PHPUnit-test-running-order-management) while the official docs are updated.|

# Inspiration
Are your unit tests truly independent units?

In the past I have often been bitten by test suites that do not test what they claim to test. A common reason is hidden dependencies in the runtime context. Worst case this has caused applications to fail when deployed in production environments.

Changing the order of unit tests turns out to be a simple and effective way of bringing these assumptions out in the open.

Thanks to @crazybus and @geogdog for providing real-world scenarios and feedback during development.

# Summary of functionality
The test ordering functionality can be configured using command line parameters:
- `--reverse-order` run tests in *reverse order*
- `--random-order` run tests in *random order*
- `--random-order-seed=<number>` fix the random order, allowing random runs to be reproduced
- `--ignore-dependencies` disable the bundled dependency resolver

# Design decisions
## Low impact, high performance
The feature will only wake up when there is work to do. The iterative dependency resolver is designed to do as little work as possible, using few iterations and minimal memory. It uses `@depends` annotations to keep as many dependent tests as possible in the required order.

No unnecessary extra output is generated. Only the randomizer outputs the *random seed* using the default results printer.

## Clean abstraction
The reordering is built into the main test runner and only changes the order of the tests of each test suite. Other parts of PHPUnit are not affected, with the exception of a bit of configuration handling.

## Changes to the built-in testsuite
- added a fourth `MultiDependencyTest` and updated existing tests to match
- added new TextUI tests for `--reverse-order`, `--random-order` and `--ignore-dependencies`
- added `@depends` for testGlobalsBackupPre() to PHPUnit's own testsuite

# Future work
- update the main PHPUnit documentation to reflect the changes in (optional) order and dependency handling
- add configuration options for `phpunit.xml` in addition to the command line flags
- add options and/or annotations for finer grained reordering config (e.g. "never reorder this suite")

### CODE REVIEW

File: tests/Util/ConfigurationTest.php [hunk 1]
```diff
@@ -330,11 +329,12 @@ public function testPHPConfigurationIsHandledCorrectly(): void
...
     {
-        $_ENV['foo'] = false;
+        $fooValue    = \uniqid();
+        $_ENV['foo'] = $fooValue;
         $this->configuration->handlePHPConfiguration();
 
-        $this->assertFalse($_ENV['foo']);
-        $this->assertEquals(1, \getenv('foo'));
+        $this->assertEquals($fooValue, $_ENV['foo']);
+//        $this->assertEquals(false, \getenv('foo'));
```
Comment (sebastianbergmann) [inline tests/Util/ConfigurationTest.php]: Is this commented out intentionally?

File: src/Framework/TestCase.php [hunk 2]
```diff
@@ -1002,6 +1017,49 @@ public function setPreserveGlobalState(bool $preserveGlobalState): void
         $this->preserveGlobalState = $preserveGlobalState;
     }
 
+    /**
+     * @param string $order
+     */
+    public function setTestRunningOrder($order): void
```
Comment (sebastianbergmann) [inline src/Framework/TestCase.php]: Please use scalar type declarations when possible (like here). Also please do not use `@param` annotations etc. when they provide no additional information.

File: src/Framework/TestSuiteIterator.php [hunk 3]
```diff
@@ -91,4 +123,72 @@ public function hasChildren(): bool
     {
         return $this->tests[$this->position] instanceof TestSuite;
     }
+
+    /**
+     * Reorder Tests within a TestCase in such a way as to resolve as many dependencies as possible.
+     * The algorithm will leave the tests in original running order when it can.
```
Comment (sebastianbergmann) [inline src/Framework/TestSuiteIterator.php]: What about cross-class dependencies? They are possible, though I rarely see them "in the wild".

**Comment by sebastianbergmann** [comment]
Thank you, @epdenouden, for working on this. I will look into this more deeply as soon as I can. I hope we can get this into PHPUnit 7.2.

File: src/Framework/TestCase.php (hunk 2 above)
Comment (epdenouden) [reply src/Framework/TestCase.php]: Fixed

**Comment by sebastianbergmann** [comment]
@epdenouden I have moved the current state of this to a new branch: https://github.com/sebastianbergmann/phpunit/tree/test-reorder.

Please rebase this pull request onto that branch and target that branch for future changes. Thanks!

**Comment by sebastianbergmann** [comment]
I am wondering whether it would make sense to move the reorder logic from `TestSuiteIterator` into separate classes, one for each type of reorder (random, dependencies, ...).

Pseudo Code:
```php
<?php
class DependencyResolver extends ArrayIterator
{
    public function __construct(TestSuiteIterator $tests)
    {
        $tests = iterator_to_array($tests);

        // reorder $tests

        parent::__construct($tests);
    }
}
```

**Comment by sebastianbergmann** [comment]
I am wondering whether it would make sense to add an extension hook for test reordering. This would make @padraic and everyone interested in #2660 happy.

**Comment by sebastianbergmann** [comment]
@lstrojny This is probably interesting to you, too, considering "clever and smart".

**Comment by epdenouden** [comment]
@sebastianbergmann Thanks for the quick replies! I will rebase and continue making the basic version clean and stable for 7.2. Good ideas for making this cleaner and more extensible. I am up for doing more coding for that.

Re: the clever and fast logic: this project is actually the groundwork for something that we call the _Breakfast_ extension, it automatically sorts the tests to run skipped, failed, etc first, but keeps their dependencies working. It also measures test and suite timings to make better sorting predictions.

**Comment by sebastianbergmann** [comment]
The "breakfast" thing sounds interesting. Lets get this sorted first, then have a look at that. Maybe it can also be integrated into the "core", maybe we can put it into an extension that uses the new, yet-to-be-added hook.

**Comment by sebastianbergmann** [comment]
@epdenouden When tests are executed in {random|reverse} order, are dependencies expressed using `@depends` taken into account? My guess is they are not (but should be) and this is why those two tests of PHPUnit's own test suite cause problems when run in reverse order.

**Comment by epdenouden** [comment]
Moinmoin @sebastianbergmann ! Yes, the `@depends` annotation should be kept in mind, they are 'food' for the dependency resolver. I am looking at the getenv() vs $_ENV issue at the moment, there the resolver becomes very visible by reordering `TestCaseTest::testStaticAttributesBackupPre()` to run before `testStaticAttributesBackupPost` and `testGlobalsBackupPost`

**Comment by epdenouden** [comment]
For example, running it like

```
./phpunit --reverse-order
[...output snipped...]
4) PHPUnit\Framework\TestCaseTest::testStaticAttributesBackupPost
This test depends on "PHPUnit\Framework\TestCaseTest::testStaticAttributesBackupPre" to pass.

/Users/ewout/proj/phpunit-new-order/src/Framework/TestCase.php:1795
/Users/ewout/proj/phpunit-new-order/src/Framework/TestCase.php:695
/Users/ewout/proj/phpunit-new-order/src/Framework/TestSuite.php:788
/Users/ewout/proj/phpunit-new-order/src/Framework/TestSuite.php:788
/Users/ewout/proj/phpunit-new-order/src/Framework/TestSuite.php:788
/Users/ewout/proj/phpunit-new-order/src/TextUI/TestRunner.php:565
/Users/ewout/proj/phpunit-new-order/src/TextUI/Command.php:202
/Users/ewout/proj/phpunit-new-order/src/TextUI/Command.php:155

5) PHPUnit\Framework\TestCaseTest::testGlobalsBackupPost
This test depends on "PHPUnit\Framework\TestCaseTest::testGlobalsBackupPre" to pass.

/Users/ewout/proj/phpunit-new-order/src/Framework/TestCase.php:1795
/Users/ewout/proj/phpunit-new-order/src/Framework/TestCase.php:695
/Users/ewout/proj/phpunit-new-order/src/Framework/TestSuite.php:788
/Users/ewout/proj/phpunit-new-order/src/Framework/TestSuite.php:788
/Users/ewout/proj/phpunit-new-order/src/Framework/TestSuite.php:788
/Users/ewout/proj/phpunit-new-order/src/TextUI/TestRunner.php:565
/Users/ewout/proj/phpunit-new-order/src/TextUI/Command.php:202
/Users/ewout/proj/phpunit-new-order/src/TextUI/Command.php:155

FAILURES!
Tests: 1657, Assertions: 2898, Failures: 2, Skipped: 5.
```

whereas

```
./phpunit --reverse-order --reorder-dependencies
[...output snipped...]
FAILURES!
Tests: 1657, Assertions: 2912, Failures: 2, Skipped: 3.
```

**Comment by epdenouden** [comment]
@sebastianbergmann The TestSuiteSorter looks nice and tidy like this :)

**Comment by sebastianbergmann** [comment]
* `TestSuiteIterator` must not know anything about sorting/reordering, therefore it must not know `TestSuiteSorter`
* `TestSuite` does not need to know anything about sorting/reordering once `TestSuiteIterator` does not know `TestSuiteSorter` anymore
* Tests are sorted (in the runner) before the first test is executed
* This is also the point where we can let extensions hook into the sorting/reordering

**Comment by epdenouden** [comment]
@sebastianbergmann Excellent points, thanks for taking the time to review and even cleanup the state of things overnight. I'll rebase and continue where I left off yesterday evening.

**Comment by epdenouden** [comment]
@sebastianbergmann Fresh look this morning, bit of clean up and done some more testing and performance timings. Looks stable and efficient. Anything missing to target [milestone 7.2](https://github.com/sebastianbergmann/phpunit/milestone/24)?

It has been a good learning experiment. Thanks for the all the time you spend on building and maintaining one of my main tools. :-)

**Comment by epdenouden** [comment]
I'll focus on the original run-failed-and-skipped-first project for which this was the first building block. For Breakfast I'll look through my notes for ideas that have been left by the wayside, such as config file parameters, optional extra output (for running order, timings, random seed in other formats, etc etc). It can now simply become an extra sorting-method.

**Comment by sebastianbergmann** [comment]
@epdenouden Thank you for your ongoing work on this. I plan to look at (and hopefully merge) this over the next weekend.

**Comment by epdenouden** [comment]
@sebastianbergmann Thanks! Will check my notes for any open issues before then. I've put together a description and examples to try to see the feature at work: https://github.com/epdenouden/phpunit/wiki/PHPUnit-test-running-order-management

**Comment by sebastianbergmann** [comment]
This branch cannot be rebased due to conflicts. Can you please fix the conflicts? Thanks!

**Comment by epdenouden** [comment]
Yes, will do that when I get back from lunch

**Comment by epdenouden** [comment]
Github's error message isn't very helpful unfortunately, what conflicts? :) When comparing my `new-order` branch to both your `test-reorder` and `master` Github seems happy to do it. For `test-reorder` I got this message:

image

I did see two conflicts when merging `7.1` in outside this PR it seems, one in the PHPStan config and one `Runner/Version.php`. Hope I have fixed them correctly; running a few test scenarios and then I'll push the merge changes from 7.1.5 here.

**Comment by epdenouden** [comment]
Okidoki. I'll be heading for some wifi and coffee while the checks are running :)

**Comment by sebastianbergmann** [comment]
Strange, GitHub still tells me "This branch cannot be rebased due to conflicts". I was able to rebase it manually locally, though.

**Comment by epdenouden** [comment]
Not sure what to do about the Travis failing build. The change to PHPStan level 2 in f59852f593cdcce49e4b799c2a5b73233dacfc01 causes it to fail.

**Comment by epdenouden** [comment]
Found the missing piece of PHPStan configuration that went missing while merging a25126a852f8476b5674e417bb2bd1e013d4b4af. PHPStan works as before.

File: src/Framework/TestSuiteIterator.php [hunk 4]
```diff
@@ -14,21 +14,18 @@
 /**
  * Iterator for test suites.
  */
-class TestSuiteIterator implements RecursiveIterator
+final class TestSuiteIterator implements RecursiveIterator
```
Comment (keradus) [inline src/Framework/TestSuiteIterator.php]: btw, this is a  BC breaker for all of us that were using this class. (class is not marked as `@internal`)

File: src/Util/TestDox/TestResult.php [hunk 5]
```diff
@@ -9,147 +9,19 @@
  */
 namespace PHPUnit\Util\TestDox;
 
-final class TestResult
```
Comment (keradus) [inline src/Util/TestDox/TestResult.php]: btw, this is a BC breaker for all of us that were using this class. (class is not marked as `@internal`)
making it non-class (->interface) is even bigger change

**Review by keradus — COMMENTED** [review]
great work in general!

let me just point out 2 bc breakes for edge-case users

**Comment by sebastianbergmann** [comment]
@keradus PHPUnit still lacks a clear statement of what is considered part of its public API. I hope to have this in place "soon", at the latest in time for PHPUnit 8.

Off the top of my head, I would say that everything internal with the exception of `Test`, `TestCase`, `TestListener`, etc.

**Comment by keradus** [comment]
I know there is no clear statement on that, but especially as it's not documented, all is public and non-internal :(

File: src/Framework/TestSuiteIterator.php (hunk 4 above)
Comment (epdenouden) [reply src/Framework/TestSuiteIterator.php]: In the end I didn't need to change many of the existing internals as the code moved to TestSuiteSorter. Extensions support is coming, I tried that first :)

**Comment by keradus** [comment]
just wondering... can we make `random-order` a default one please?
or at least, allow to set it via config file

**Comment by sebastianbergmann** [comment]
-1 for default (because of `--testdox`, for instance)
+0 for configuration

Come to think of it, we should probably disable the dependency resolution by default in PHPUnit 7.2 and turn it on by default in PHPUnit 8.

**Comment by keradus** [comment]
value of this PR is nice, yet requires manual work that one always remember to use that flag.
with config file, it would be way easier

**Comment by sebastianbergmann** [comment]
@keradus I have opened #3131 for this and will (hopefully) implement it tomorrow.

**Comment by keradus** [comment]
Thanks for accepting the request :+1:
no need to treat this one as prio, no rush on it ;)

**Comment by epdenouden** [comment]
Indeed thanks for merging it in. I will continue on the followup feature for rerunning failures first now

**Comment by epdenouden** [comment]
@keradus having reordering as a default would be fun, but it would spook a lot of people that run phpunit in "any 7.x" version will do. Since testing frameworks are important for creating trust that is not so great :)

**Comment by keradus** [comment]
> Indeed thanks for merging it in. I will continue on the followup feature for rerunning failures first now

Sounds like https://github.com/fiunchinho/phpunit-randomizer is obsolete, now it's time for
https://github.com/lstrojny/phpunit-clever-and-smart ? :)

> having reordering as a default ...

That's why I suggested to have at least config option, thus one don't need to remember about option switch every single time.
On the other hand, if I have tests that are always passing, and then we would detect that `assertSame` doesnt work for `string` for some weird reason, I do expect `assertSame` to be fixed and yes, my tests would start failing. It is not that my test was working and suddenly it stopped. It's that my tests were not able to detect real issue in my application, and after update of testing framework they are. And it's super cool to know that my src has bugs.

**Pull request closed without merging**